mathics/builtin/lists.py

```diff
@@ -2096,14 +2099,234 @@ def apply(self, list, sep, evaluation):
+    # in its own bin), we expect to make O(n) comparisons (if the hash function
+    # does not fail us by distributing items very unevenly).
+
+    # IMPORTANT NOTE ON ATOM'S HASH FUNCTIONS / this code relies on this assumption:
+    #
+    # if SameQ[a, b] == true then hash(a) == hash(b)
+    #
+    # more specifically, this code bins items based on their hash code, and only if
+    # the hash code matches, is SameQ evoked.
+    #
+    # this assumption has been checked for these types: Integer, Real, Complex,
+    # String, Rational (*), Expression, Image; new atoms need proper hash functions
+    #
+    # (*) Rational values are sympy Rationals which are always held in reduced form
+    # and thus are hashed correctly (see sympy/core/numbers.py:Rational.__eq__()).
+
+    def __init__(self):
+        self._hashes = defaultdict(list)
+
+    def select(self):
+        hashes = self._hashes
+        return lambda elem: hashes[hash(elem)]
+
+    def same(self):
+        return lambda a, b: a.same(b)
+
+
+class _GatherBin:
+    def __init__(self, item):
+        self._items = [item]
+        self.add_to = self._items.append
+
+    def from_python(self):
+        return Expression('List', *self._items)
+
+
+class _TallyBin:
+    def __init__(self, item):
+        self._item = item
+        self._count = 1
+
+    def add_to(self, item):
+        self._count += 1
+
+    def from_python(self):
+        return Expression('List', self._item, Integer(self._count))
+
+
+class _DeleteDuplicatesBin:
+    def __init__(self, item):
+        self._item = item
+        self.add_to = lambda elem: None
+
+    def from_python(self):
+        return self._item
+
+
+class _GatherOperation(Builtin):
+    rules = {
+        '%(name)s[list_]': '%(name)s[list, SameQ]'
+    }
+
+    messages = {
+        'normal': 'Nonatomic expression expected at position `1` in `2`.',
+        'list': 'List expected at position `2` in `1`.',
+        'smtst': ("Application of the SameTest yielded `1`, which evaluates "
+                  "to `2`. The SameTest must evaluate to True or False at "
+                  "every pair of elements."),
+    }
+
+    def apply(self, list, test, evaluation):
+        '%(name)s[list_, test_]'
+
+        if list.is_atom():
+            expr = Expression(self.get_name(), list, test)
+            return evaluation.message(self.get_name(), 'normal', 1, expr)
+
+        if list.get_head_name() != 'System`List':
+            expr = Expression(self.get_name(), list, test)
+            return evaluation.message(self.get_name(), 'list', expr, 1)
+
+        if _is_sameq(test):
+            return self._gather(list, _FastEquivalence())
+        else:
+            return self._gather(list, _SlowEquivalence(test, evaluation, self.get_name()))
+
+    def _gather(self, a_list, equivalence):
+        bins = []
+        Bin = self._bin
+
+        select = equivalence.select()
+        same = equivalence.same()
```

Comment:
Unless I'm missing something, the usage of `.select()` and `.same()` here seems more complicated than it needs to be. I don't think these need to return closures here.

How about refactoring these methods so this function can just call `equivalance.select(elem)` and `equivalence.same(prototype, elem)`? This would also allow `_make_test_pair()` to be folded into `_SlowEquivalence.same()`.